neon_data_models/models/api/llm.py

```diff
@@ -0,0 +1,194 @@
+        # If beams are specified, make sure valid `stream` and `beam_search`
+        # values are specified
+        if self.best_of > 1:
+            if self.stream is True:
+                raise ValueError("Cannot stream with a `best_of` value "
+                                 "greater than 1")
+            if self.beam_search is False:
+                raise ValueError("Cannot have a `best_of` value other than 1 "
+                                 "if `beam_search` is False")
+            self.stream = False
+            self.beam_search = True
+        # If streaming, beam_search must be False
+        if self.stream is True:
+            if self.beam_search is True:
+                raise ValueError("Cannot enable both `stream` and "
+                                 "`beam_search`")
+            self.beam_search = False
+        # If beam search is enabled, streaming must be False
+        if self.beam_search is True:
+            if self.stream is True:
+                raise ValueError("Cannot enable both `stream` and "
+                                 "`beam_search`")
+            self.stream = False
+        if self.stream is None and self.beam_search is None:
+            self.stream = True
+            self.beam_search = False
+
+        assert isinstance(self.stream, bool)
+        assert isinstance(self.beam_search, bool)
+
+        # If beam search is enabled, temperature must be set to 0.0
+        if self.beam_search:
+            assert self.temperature == 0.0
+        return self
+
+    @property
+    def messages(self) -> List[dict]:
+        """
+        Get chat history as a list of dict messages
+        """
+        return [{"role": m[0], "content": m[1]} for m in self.history]
+
+    def to_completion_kwargs(self, mq2role: dict = None) -> dict:
+        """
+        Get kwargs to pass to an OpenAI completion request.
+        @param mq2role: dict mapping `llm` and `user` keys to `role` values to
+            use in message history.
+        """
+        mq2role = mq2role or _DEFAULT_MQ_TO_ROLE
+        history = self.messages[-2*self.max_history:]
+        for msg in history:
+            msg["role"] = mq2role.get(msg["role"]) or msg["role"]
+        history.insert(0, {"role": "system",
+                           "content": self.persona.system_prompt})
+        return {"model": self.model,
+                "messages": history,
+                "max_tokens": self.max_tokens,
+                "temperature": self.temperature,
+                "stream": self.stream,
+                "extra_body": {"add_special_tokens": True,
+                               "repetition_penalty": self.repetition_penalty,
+                               "use_beam_search": self.beam_search,
+                               "best_of": self.best_of}}
+
+
+class LLMResponse(BaseModel):
+    response: str = Field(description="LLM Response to the input query")
+    history: List[Tuple[Literal["user", "llm"], str]] = Field(
```

Comment:
Here we can also return stop reason
Was it because it reached `self.max_tokens` or generated `EOS`

In OpenAI API it's called `finish_reason: Literal["length", "stop"]`, [reference](https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format)